# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import numpy as np # importing numpy for numerical computation
import random
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric
from sklearn.tree import DecisionTreeRegressor
from tqdm import tqdm

In [3]:
boston = load_boston()
x = boston.data #independent variables
y = boston.target #target variable

In [4]:
print(x.shape)

(506, 13)


In [5]:
print(x[:5])

[[6.3200e-03 1.8000e+01 2.3100e+00 0.0000e+00 5.3800e-01 6.5750e+00
  6.5200e+01 4.0900e+00 1.0000e+00 2.9600e+02 1.5300e+01 3.9690e+02
  4.9800e+00]
 [2.7310e-02 0.0000e+00 7.0700e+00 0.0000e+00 4.6900e-01 6.4210e+00
  7.8900e+01 4.9671e+00 2.0000e+00 2.4200e+02 1.7800e+01 3.9690e+02
  9.1400e+00]
 [2.7290e-02 0.0000e+00 7.0700e+00 0.0000e+00 4.6900e-01 7.1850e+00
  6.1100e+01 4.9671e+00 2.0000e+00 2.4200e+02 1.7800e+01 3.9283e+02
  4.0300e+00]
 [3.2370e-02 0.0000e+00 2.1800e+00 0.0000e+00 4.5800e-01 6.9980e+00
  4.5800e+01 6.0622e+00 3.0000e+00 2.2200e+02 1.8700e+01 3.9463e+02
  2.9400e+00]
 [6.9050e-02 0.0000e+00 2.1800e+00 0.0000e+00 4.5800e-01 7.1470e+00
  5.4200e+01 6.0622e+00 3.0000e+00 2.2200e+02 1.8700e+01 3.9690e+02
  5.3300e+00]]


## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

* <font color='red'><b> Note - While selecting the random 60% datapoints from the whole data, make sure that the selected datapoints are all exclusive, repetition is not allowed. </b></font>

<font color='red'><b>Step - 2 </b></font>

* <font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>
    *  Build a regression trees on each of 30 samples.
    *  Computed the predicted values of each data point(506 data points) in your corpus.
    *  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
    *  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>
    *  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
    *  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

    * Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
    * Predict the house price for this point as mentioned in the step 2 of Task 1.

## <font color='red'><b>A few key points</b></font>

* Remember that the datapoints used for calculating MSE score contain some datapoints that were initially used while training the base learners (the 60% sampling). This makes these datapoints partially seen (i.e. the datapoints used for calculating the MSE score are a mixture of seen and unseen data).
Whereas, the datapoints used for calculating OOB score have only the unseen data. This makes these datapoints completely unseen and therefore appropriate for testing the model's performance on unseen data.

* Given the information above, if your logic is correct, the calculated MSE score should be less than the OOB score.

* The MSE score must lie between 0 and 10.
* The OOB score must lie between 10 and 35.

* The difference between the left nad right confidence-interval values must not be more than 10. Make sure this is true for both MSE and OOB confidence-interval values.

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/OfcFrUP.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [6]:
def generating_samples(input_data, target_data):
    """
    In this function, we will write code for generating 30 samples.
    """
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link 
    # https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples 
    # return sampled_input_data , sampled_target_data,selected_rows,selected_columns
    # note please return as lists
    
    total_rows, total_cols = input_data.shape[0], input_data.shape[1]
    selected_row_size = int(total_rows * 0.6)
    replaced_row_size = total_rows - selected_row_size
    random_col_size = random.randint(3, 13)
    
    selecting_rows = np.random.choice(a=total_rows, size=selected_row_size, replace=False)
    replacing_rows = np.random.choice(a=selecting_rows, size=replaced_row_size, replace=True)
    
    selecting_columns = np.random.choice(a=total_cols, size=random_col_size, replace=False)
    
    sample_data = input_data[selecting_rows[:, None], selecting_columns]
    target_sample_data = target_data[selecting_rows]
    
    replicated_sample_data = input_data[replacing_rows[:, None], selecting_columns]
    target_replicated_sample_data = target_data[replacing_rows]
    
    final_sample_data = np.vstack(tup=(sample_data, replicated_sample_data))
    final_target_data = np.vstack(tup=(target_sample_data.reshape(-1, 1),
                                       target_replicated_sample_data.reshape(-1, 1)))
    return final_sample_data, final_target_data, selecting_rows, selecting_columns

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [7]:
def grader_samples(a, b, c, d):
    length = (len(a) == 506  and len(b) == 506)
    sampled = (len(a) - len(set([str(i) for i in a])) == 203)
    rows_length = (len(c) == 303)
    column_length = (len(d) >= 3)
    assert(length and sampled and rows_length and column_length)
    return True
a, b, c, d = generating_samples(input_data=x, target_data=y)
grader_samples(a=a, b=b, c=c, d=d)

True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [8]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list

def create_the_samples(input_data, target_data, samples=30):
    """
    This function creates the samples.
    """
    list_input_data = list()
    list_output_data = list()
    list_selected_row = list()
    list_selected_columns = list()
    for i in range(0, samples):
        a, b, c, d = generating_samples(input_data=input_data, target_data=target_data)
        list_input_data.append(a)
        list_output_data.append(b)
        list_selected_row.append(c)
        list_selected_columns.append(d)
    return (list_input_data,
            list_output_data,
            list_selected_row,
            list_selected_columns)

In [9]:
(list_input_data,
 list_output_data,
 list_selected_row,
 list_selected_columns) = create_the_samples(input_data=x, target_data=y)

<font color='cyan'> <b>Grader function - 2 </b></font>

In [10]:
def grader_30(a):
    assert(len(a) == 30 and len(a[0]) == 506)
    return True
grader_30(a=list_input_data)

True

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [11]:
def fit_decision_regressor(list_input_data, list_output_data):
    """
    This functions fits the decision regressors.
    """
    list_of_all_models = list()
    for input_x, target_y in zip(list_input_data, list_output_data):
        model_i = DecisionTreeRegressor(max_depth=None)
        model_i.fit(X=input_x, y=target_y)
        list_of_all_models.append(model_i)
    return list_of_all_models

In [12]:
list_of_all_models = fit_decision_regressor(list_input_data=list_input_data,
                                            list_output_data=list_output_data)

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [13]:
def compute_mse(input_data, target_data, list_input_data, list_selected_columns, list_of_all_models):
    """
    This function computes mean squared error for each data point using all the model_i.
    """
    y_pred = list()
    for data_point in input_data:
        data_point_pred = list()
        for i, model_i in enumerate(list_of_all_models):
            Xi = data_point[list_selected_columns[i]]
            y_cap = model_i.predict(X=Xi.reshape(1, -1))
            data_point_pred.append(y_cap)
        y_pred.append(np.mean(data_point_pred))
    y_pred = np.array(y_pred)
    mse = mean_squared_error(y_true=target_data, y_pred=y_pred)
    return y_pred, mse

In [14]:
_, mse = compute_mse(input_data=x,
                     target_data=y,
                     list_input_data=list_input_data,
                     list_selected_columns=list_selected_columns,
                     list_of_all_models=list_of_all_models)
print(mse)

2.3213951877836325


<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $\text{OOB Score} =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [15]:
def compute_oob(input_data, target_data, list_input_data, list_selected_columns, list_of_all_models):
    """
    This function computes mean squared error for each data point using all the model_i.
    """
    y_pred = list()
    for data_point in input_data:
        data_point_pred = list()
        for i, model_i in enumerate(list_of_all_models):
            Xi = data_point[list_selected_columns[i]]
            if not any(np.equal(Xi, list_input_data[i]).all(1)):
                y_cap = model_i.predict(X=Xi.reshape(1, -1))
                data_point_pred.append(y_cap)
        y_pred.append(np.mean(data_point_pred))
    y_pred = np.array(y_pred)
    mse = mean_squared_error(y_true=target_data, y_pred=y_pred)
    return y_pred, mse

In [16]:
_, oob = compute_oob(input_data=x,
                     target_data=y,
                     list_input_data=list_input_data,
                     list_selected_columns=list_selected_columns,
                     list_of_all_models=list_of_all_models)
print(oob)

14.793952404025633


# <font color='blue'><b>Task 2</b></font>

In [17]:
def get_mse_oob_scores(input_data, target_data, times=35):
    """
    This function gets the mse and oob scores.
    """
    mse_scores = list()
    oob_scores = list()
    for t in tqdm(range(times)):
        (list_input_data,
         list_output_data,
         list_selected_row,
         list_selected_columns) = create_the_samples(input_data=x, target_data=y)
        list_of_all_models = fit_decision_regressor(list_input_data=list_input_data,
                                                    list_output_data=list_output_data)
        _, mse = compute_mse(input_data=input_data,
                             target_data=target_data,
                             list_input_data=list_input_data,
                             list_selected_columns=list_selected_columns,
                             list_of_all_models=list_of_all_models)
        _, oob = compute_oob(input_data=input_data,
                             target_data=target_data,
                             list_input_data=list_input_data,
                             list_selected_columns=list_selected_columns,
                             list_of_all_models=list_of_all_models)
        mse_scores.append(mse)
        oob_scores.append(oob)
    return mse_scores, oob_scores

In [18]:
mse_scores, oob_scores = get_mse_oob_scores(input_data=x, target_data=y)

100%|███████████████████████████████████████████████████████████████████| 35/35 [00:33<00:00,  1.05it/s]


In [19]:
def compute_ci(scores):
    """
    This function computes the Confidence Interval.
    """
    scores = np.array(scores)
    n = len(scores)
    sample_mean = np.mean(a=scores)
    sample_std = np.sqrt(np.sum((scores - sample_mean) ** 2) / (n - 1))
    std_error = sample_std / np.sqrt(n)
    lower = sample_mean - (2 * std_error)
    upper = sample_mean + (2 * std_error)
    return [lower, upper]

In [20]:
mse_ci = compute_ci(scores=mse_scores)
print("Confidence Interval for MSE Scores → {}".format(mse_ci))

Confidence Interval for MSE Scores → [2.2787120038900204, 2.47229466684713]


In [21]:
oob_ci = compute_ci(scores=oob_scores)
print("Confidence Interval for OOB Scores → {}".format(oob_ci))

Confidence Interval for OOB Scores → [13.151759318491756, 14.088865314856918]


# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [22]:
def predict_xq(X_q, list_input_data, list_selected_columns, list_of_all_models):
    """
    This function predicts the query point.
    """
    X_q = np.array(X_q)
    preds = list()
    for i, model_i in enumerate(list_of_all_models):
        X_q_by_feature = X_q[list_selected_columns[i]]
        p = model_i.predict(X=X_q_by_feature.reshape(1, -1))
        preds.append(p)
    preds = np.array(preds)
    prediction = np.mean(preds)
    return prediction

In [23]:
X_q = [0.18, 20.0, 5.00, 0.0, 0.421, 5.60, 72.2, 7.95, 7.0, 30.0, 19.1, 372.13, 18.60]
prediction = predict_xq(X_q=X_q,
                        list_input_data=list_input_data,
                        list_selected_columns=list_selected_columns,
                        list_of_all_models=list_of_all_models)
print(prediction)

19.863333333333333


<font color='red'><b>Write observations for task 1, task 2, task 3 in detail.</b></font>

In this assignment, I have learned how bagging is implemented. I am clear with the concept. The flow charts and pseudo codes are really of great help.

The size of the data is very small. Hence, the training was very faster.

- In task 1, I learned the process of sampling with replacement and constructing the Random Forest with dense regressors.

- In task 2, I learned computing the CI. The MSE for training data is 2.321 and is within the confidence interval → [2.278, 2.472].

- In task 3, I learned how to predict a new query point. And this is very useful lesson.

End of the file.